### PR TITLE
Add package discover refresh when component/trait definition are registered

### PR DESCRIPTION
**Pull request by yangsoon**
fix https://github.com/oam-dev/kubevela/issues/1387

refresh package discover when component/trait definition are registered.

- [x] add discover refresh
- [x] add test

### CODE REVIEW

File: pkg/oam/util/helper.go [hunk 1]
```diff
@@ -97,6 +97,9 @@ const (
 
 	// ErrCreateConvertedWorklaodDefinition is the error while apply a WorkloadDefinition
 	ErrCreateConvertedWorklaodDefinition = "cannot create converted WorkloadDefinition %s: %v"
+
+	// ErrRefreshPackageDiscover is the error while refresh PackageDiscover
+	ErrRefreshPackageDiscover = "cannot refresh PackageDiscover : %v"
```
Comment (wonderflow) [inline pkg/oam/util/helper.go:102]:
```suggestion
	ErrRefreshPackageDiscover = "cannot discover the open api of the CRD: %v"
```

File: pkg/controller/core.oam.dev/v1alpha2/core/traits/traitdefinition/testdata/example.go [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+var (
+	FooKind             = reflect.TypeOf(Foo{}).Name()
+	FooGroupKind        = schema.GroupKind{Group: Group, Kind: FooKind}.String()
+	FooKindAPIVersion   = FooGroupKind + "." + SchemeGroupVersion.String()
+	FooGroupVersionKind = SchemeGroupVersion.WithKind(FooGroupKind)
+)
+
+func init() {
+	SchemeBuilder.Register(&Foo{}, &FooList{})
+}
```
Comment (wonderflow) [inline pkg/controller/core.oam.dev/v1alpha2/core/traits/traitdefinition/testdata/example.go:79]: we don't need to add such a structure just for test.

Comment (yangsoon) [reply pkg/controller/core.oam.dev/v1alpha2/core/traits/traitdefinition/testdata/example.go:79]: got it

File: pkg/controller/core.oam.dev/v1alpha2/core/traits/traitdefinition/traitdefinition_suite_test.go [hunk 3]
```diff
@@ -51,6 +54,7 @@ var _ = BeforeSuite(func(done Done) {
 	useExistCluster := false
 	testEnv = &envtest.Environment{
 		CRDDirectoryPaths: []string{
+			"./testdata/crd",
```
Comment (wonderflow) [inline pkg/controller/core.oam.dev/v1alpha2/core/traits/traitdefinition/traitdefinition_suite_test.go:57]: unused line

Comment (yangsoon) [reply pkg/controller/core.oam.dev/v1alpha2/core/traits/traitdefinition/traitdefinition_suite_test.go:57]: got it

File: pkg/controller/utils/capability.go [hunk 4]
```diff
@@ -306,8 +308,25 @@ func getOpenAPISchema(capability types.Capability) ([]byte, error) {
 }
 
 // generateOpenAPISchemaFromCapabilityParameter returns the parameter of a definition in cue.Value format
-func generateOpenAPISchemaFromCapabilityParameter(capability types.Capability) ([]byte, error) {
-	return GenerateOpenAPISchemaFromDefinition(capability.Name, capability.CueTemplate)
```
Comment (wonderflow) [inline pkg/controller/utils/capability.go:310]: GenerateOpenAPISchemaFromDefinition does this func still being used?

File: pkg/oam/discoverymapper/mapper.go [hunk 5]
```diff
@@ -55,9 +58,12 @@ func New(c *rest.Config) (DiscoveryMapper, error) {
 // GetMapper will get the cached restmapper, if nil, it will create one by refresh
 // Prefer lazy discovery, because resources created after refresh can not be found
 func (d *DefaultDiscoveryMapper) GetMapper() (meta.RESTMapper, error) {
+	d.mutex.RLock()
```
Comment (wonderflow) [inline pkg/oam/discoverymapper/mapper.go:61]:
```suggestion
	d.mutex.RLock()
	defer d.mutex.RUnlock()
```

File: pkg/oam/discoverymapper/mapper.go [hunk 6]
```diff
@@ -55,9 +58,12 @@ func New(c *rest.Config) (DiscoveryMapper, error) {
...
 	if d.mapper == nil {
+		d.mutex.RUnlock()
```
Comment (wonderflow) [inline pkg/oam/discoverymapper/mapper.go:63]:
```suggestion
```

File: pkg/oam/discoverymapper/mapper.go [hunk 7]
```diff
@@ -55,9 +58,12 @@ func New(c *rest.Config) (DiscoveryMapper, error) {
...
 		return d.Refresh()
 	}
+	d.mutex.RUnlock()
```
Comment (wonderflow) [inline pkg/oam/discoverymapper/mapper.go:66]:
```suggestion
```

File: pkg/controller/utils/capability.go (hunk 4 above)
Comment (yangsoon) [reply pkg/controller/utils/capability.go:310]: i will fix it

**Review by wonderflow — APPROVED** [review]
LGTM

**Pull request merged**